File: StoneCarveManager.Services/Database/Entities/Product.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StoneCarveManager.Services.Database.Entities
{
    public class Product
    {
        public int Id { get; set; }
        public string Name { get; set; } = string.Empty;
        public string Description { get; set; } = string.Empty;
        public decimal Price { get; set; }

        public int StockQuantity { get; set; } = 0;

        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        public DateTime? UpdatedAt { get; set; }

        public string? Dimensions { get; set; }

        // Product weight for shipping calculations (in grams)
        public decimal? Weight { get; set; }
        public int EstimatedDays { get; set; } = 7;
        public bool IsInPortfolio { get; set; } = true;
        public int ViewCount { get; set; } = 0;

        // Category relationship
        public int? CategoryId { get; set; }
        public Category? Category { get; set; }
        
        // Material relationship
        public int? MaterialId { get; set; }
        public Material? Material { get; set; }

        // Navigation property for assets (images)
        public ICollection<ProductImage> Images { get; set; } = new List<ProductImage>();


        // Navigation property for product reviews
        public ICollection<ProductReview> Reviews { get; set; } = new List<ProductReview>();

        // Navigation property for order items
        public ICollection<OrderItem> OrderItems { get; set; } = new List<OrderItem>();

        // Navigation property for cart items
        public ICollection<CartItem> CartItems { get; set; } = new List<CartItem>();
        public string ProductState { get; set; } = "draft";

        // NEW: Portfolio-specific fields
        [StringLength(4000)]
        public string? PortfolioDescription { get; set; }  // Detailed project description

        [StringLength(2000)]
        public string? ClientChallenge { get; set; }  // "The Challenge" section

        [StringLength(2000)]
        public string? OurSolution { get; set; }  // "Our Solution" section

        [StringLength(2000)]
        public string? ProjectOutcome { get; set; }  // "The Outcome" section

        [StringLength(200)]
        public string? Location { get; set; }  // Project location (e.g., "Park Mostar")

        public int? CompletionYear { get; set; }  // Year project was completed

        public int? ProjectDuration { get; set; }  // How many days it took

        [StringLength(500)]
        public string? TechniquesUsed { get; set; }  // "Hand-carved, CNC, 
    }
}

File: UI/stonecarve_manager_desktop/lib/models/product.dart
import 'category.dart';
import 'material.dart' as stone_material;

class Product {
  int? id;
  String? name;
  String? description;
  double? price;
  int? stockQuantity;
  DateTime? createdAt;
  DateTime? updatedAt;
  String? dimensions;
  double? weight;
  int? estimatedDays;
  bool? isInPortfolio;
  int? viewCount;
  int? categoryId;
  String? categoryName;
  int? materialId;
  String? materialName;
  String? productState;
  int? reviewCount;
  double? averageRating;

  // Portfolio-specific fields
  String? portfolioDescription;
  String? clientChallenge;
  String? ourSolution;
  String? projectOutcome;
  String? location;
  int? completionYear;
  int? projectDuration;
  String? techniquesUsed;

  // Related objects
  Category? category;
  stone_material.StoneMaterial? material;
  List<ProductImage>? images;
  List<ProductReview>? reviews;

  Product({
    this.id,
    this.name,
    this.description,
    this.price,
    this.stockQuantity,
    this.createdAt,
    this.updatedAt,
    this.dimensions,
    this.weight,
    this.estimatedDays,
    this.isInPortfolio,
    this.viewCount,
    this.categoryId,
    this.categoryName,
    this.materialId,
    this.materialName,
    this.productState,
    this.reviewCount,
    this.averageRating,
    this.portfolioDescription,
    this.clientChallenge,
    this.ourSolution,
    this.projectOutcome,
    this.location,
    this.completionYear,
    this.projectDuration,
    this.techniquesUsed,
    this.category,
    this.material,
    this.images,
    this.reviews,
  });

  Product.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    name = json['name'];
    description = json['description'];
    price = json['price']?.toDouble();
    stockQuantity = json['stockQuantity'];
    createdAt = json['createdAt'] != null
        ? DateTime.parse(json['createdAt'])
        : null;
    updatedAt = json['updatedAt'] != null
        ? DateTime.parse(json['updatedAt'])
        : null;
    dimensions = json['dimensions'];
    weight = json['weight']?.toDouble();
    estimatedDays = json['estimatedDays'];
    isInPortfolio = json['isInPortfolio'];
    viewCount = json['viewCount'];
    categoryId = json['categoryId'];
    categoryName = json['categoryName'];
    materialId = json['materialId'];
    materialName = json['materialName'];
    productState = json['productState'];
    reviewCount = json['reviewCount'];
    averageRating = json['averageRating']?.toDouble();

    portfolioDescription = json['portfolioDescription'];
    clientChallenge = json['clientChallenge'];
    ourSolution = json['ourSolution'];
    projectOutcome = json['projectOutcome'];
    location = json['location'];
    completionYear = json['completionYear'];
    projectDuration = json['projectDuration'];
    techniquesUsed = json['techniquesUsed'];

    if (json['category'] != null) {
      category = Category.fromJson(json['category']);
    }
    if (json['material'] != null) {
      material = stone_material.StoneMaterial.fromJson(json['material']);
    }
    if (json['images'] != null) {
      images = <ProductImage>[];
      json['images'].forEach((v) {
        images!.add(ProductImage.fromJson(v));
      });
    }
    if (json['reviews'] != null) {
      reviews = <ProductReview>[];
      json['reviews'].forEach((v) {
        reviews!.add(ProductReview.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'price': price,
      'stockQuantity': stockQuantity,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
      'dimensions': dimensions,
      'weight': weight,
      'estimatedDays': estimatedDays,
      'isInPortfolio': isInPortfolio,
      'viewCount': viewCount,
      'categoryId': categoryId,
      'categoryName': categoryName,
      'materialId': materialId,
      'materialName': materialName,
      'productState': productState,
      'reviewCount': reviewCount,
      'averageRating': averageRating,
      'portfolioDescription': portfolioDescription,
      'clientChallenge': clientChallenge,
      'ourSolution': ourSolution,
      'projectOutcome': projectOutcome,
      'location': location,
      'completionYear': completionYear,
      'projectDuration': projectDuration,
      'techniquesUsed': techniquesUsed,
      'category': category?.toJson(),
      'material': material?.toJson(),
      'images': images?.map((v) => v.toJson()).toList(),
      'reviews': reviews?.map((v) => v.toJson()).toList(),
    };
  }
}

class ProductImage {
  int? id;
  int? productId;
  String? imageUrl;
  String? altText;
  bool? isPrimary;
  DateTime? createdAt;

  ProductImage({
    this.id,
    this.productId,
    this.imageUrl,
    this.altText,
    this.isPrimary,
    this.createdAt,
  });

  ProductImage.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    productId = json['productId'];
    imageUrl = json['imageUrl'];
    altText = json['altText'];
    isPrimary = json['isPrimary'];
    createdAt = json['createdAt'] != null
        ? DateTime.parse(json['createdAt'])
        : null;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'productId': productId,
      'imageUrl': imageUrl,
      'altText': altText,
      'isPrimary': isPrimary,
      'createdAt': createdAt?.toIso8601String(),
    };
  }
}

class ProductReview {
  int? id;
  int? productId;
  int? userId;
  String? userName;
  int? rating;
  String? comment;
  DateTime? createdAt;

  ProductReview({
    this.id,
    this.productId,
    this.userId,
    this.userName,
    this.rating,
    this.comment,
    this.createdAt,
  });

  ProductReview.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    productId = json['productId'];
    userId = json['userId'];
    userName = json['userName'];
    rating = json['rating'];
    comment = json['comment'];
    createdAt = json['createdAt'] != null
        ? DateTime.parse(json['createdAt'])
        : null;
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'productId': productId,
      'userId': userId,
      'userName': userName,
      'rating': rating,
      'comment': comment,
      'createdAt': createdAt?.toIso8601String(),
    };
  }
}
